subprojects/docs/src/docs/userguide/developing-plugins/writing_in_kotlin.adoc

```diff
@@ -0,0 +1,329 @@
+
+It also describes the different development plugins to help you set up these different options.
+
+
+[[sec:dependencies]]
+== Different dependencies
+
+Like other JVM projects, Gradle plugins are jar files with dependencies. Unlike other JVM projects, Gradle provides fixed versions of some dependencies at runtime. Those dependencies are `compileOnly` dependencies and need to be compatible with the version of Gradle your plugin is running in.
+
+If you are writing convention plugins, this is usually not an issue. But if you distribute your plugins, you should pay attention to their compatibility.
+
+[[sec:gradle_api]]
+=== The Gradle API
+
+All plugins use the https://docs.gradle.org/current/javadoc/index.html[Gradle Core API] to configure a Gradle build. The Gradle API must be added as a build compile-time dependency using the helper function `gradleApi()`
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly(gradleApi())
+}
+----
+
+TIP: `gradleApi()` must be a `compileOnly` dependency, because the correct Gradle API will be provided by Gradle at runtime.
+
+==== Version of the Gradle API Version
+
+The Gradle API version provided by `gradleApi()` is the same as the version of Gradle used to build the plugin.
+
+For example, if the plugin is built using Gradle version 7.6.1, then `gradleApi()` also be version 7.6.1. If the plugin is later built with Gradle {gradleVersion}, then the `gradleApi()` will change to be version {gradleVersion}.
+
+This implicit link between affects the Gradle versions your plugin supports. It's easy to build a plugin using one version of Gradle that that has functions or classes that are not present in other versions of Gradle.
+
+This might result in runtime exceptions, such as `NoSuchMethodError`, if a plugin uses a functionality of the Gradle API that is missing in the Gradle being used to run the plugin. This might happen the plugin is run by an older Gradle version that is missing a function or class, _or_ if the plugin uses a function or class that has been removed in newer versions (which can happen when using deprecated or internal Gradle features).
+
+Supporting different versions of Gradle can be tested using https://docs.gradle.org/current/userguide/test_kit.html#sub:gradle-runner-gradle-version[TestKit].
+
+[[sec:redistributed_gradle_api]]
+=== Redistributed Gradle API
+
+To avoid the implicit link on a Gradle version that comes with using `gradleApi()`, an alternative is to use the https://docs.nokee.dev/manual/gradle-plugin-development.html[Nokee redistributed Gradle API artifacts]. These are like `gradleApi()` except they allow specifying a specific version of the Gradle API.
+
+[subs="attributes"]
+[source,kotlin]
+----
+dependencies {
+    compileOnly("dev.gradleplugins:gradle-api:{gradleVersion}")
```

Comment:
🔍 Even though the coordinate of the redistributed API is correct, the version part may sometimes differ from the exact Gradle version. The reason is that we can't overwrite the artifacts once published. If we need to patch them further, i.e. when we put the time into fixing some of the issues with redistributing the API, the plan is to release another version. We would use the version scheme `{gradleVersion}-1`, `{gradleVersion-2}`, etc. It's unclear if we should make the declaration more complicated to future proof. Instead, we recommend using the latest version of `dev.gradleplugins.gradle-plugin-development` so users can use `gradleApi("{gradleVersion}")`, which would account for any patching. The helper method works for both Groovy and Kotlin DSL. That is until Gradle releases the same API and breaks our plugin.

Reply:
Thanks for clarifying this up. I've never been hit by this so looks OK for me to leave it under the radar. Or if you have a link, we could mention it quickly and users who want more details could follow the link.

Reply:
It's a good point, and I should add a note as I see more and more projects using the coordinate directly. I opened [this issue](https://github.com/gradle-plugins/toolbox/issues/90).

Reply:
> Instead, we recommend using the latest version of dev.gradleplugins.gradle-plugin-development so users can use gradleApi("{gradleVersion}"), which would account for any patching.

How about the case when different versions of "dev.gradleplugins:gradle-api" is added as `compileOnly` dependencies to the Gradle plugin variants? I think proposed plugin will not do it?